File: Casilla.java
package proyecto;

public class Casilla {
	private boolean seguro;
	private boolean salida;
	private String colorSalida;
	private boolean puente;
	private boolean pos1;
	private boolean pos2;
	private String color1;
	private String color2;
	private int ultimo;
	
	public Casilla(boolean b, boolean b1, String s) {
		seguro = b;
		salida = b1;
		colorSalida = s;
		puente = false;
		pos1 = false;
		pos2 = false;
		color1 = new String();
		color2 = new String();
		ultimo = 0;
	}
	
	public boolean seguro() {
		return seguro;
	}
	public boolean salida() {
		return salida;
	}
	public boolean puente() {
		return puente;
	}
	public boolean pos1() {
		return pos1;
	}
	public boolean pos2() {
		return pos2;
	}
	public String color1() {
		return color1;
	}
	public String color2() {
		return color2;
	}
	public String colorSalida() {
		return colorSalida;
	}
	//Comprueba si hay movimiento posible sin matar------NO SE SI TIENE UTILIDAD
	public boolean sePuede(String s) {
		return !pos1 || (seguro && !pos2) || esSalidaSuya(s);
	}
	public boolean esValido(String s) {
		return (!pos1 || (pos1 && s!=color1));
	}
	//COmprueba si mata
	public boolean seMata(String s) {
		if(!seguro) {
			if(pos1 && s!=color1) {
				return true;
			}
		}return false;
	}
	//Comprueba si esta casilla es de salida
	public boolean esSalidaSuya(String s) {
		return salida && (s==colorSalida && (color2!=s || color1!=s));
	}
	public String introducir(String s) {
		String muerto ="NO";
		if(!pos1()) {	//Pos1 vacía
			pos1 = true;
			color1 = s;
			ultimo = 1;
			if(seguro && pos2) puente = true;
		}else if(seguro && !pos2) {	//Pos2 vacía pero tiene que ser seguro
			pos2 = true;
			color2 = s;
			puente = true;
			ultimo = 2;
		}else if(esSalidaSuya(s)) {	//Caso en el que es casilla salida
			if(ultimo==1 && s!=color1) {
				muerto = color1;
				color1 = s;
			}else if (ultimo==2 && s!=color2) {
				muerto = color2;
				color1 = s;
			}else {
				if(s!=color1) {
					muerto = color1;
					color1 = s;			
					ultimo = 1;
				}else {
					muerto = color2;
					color2 = s;
					ultimo = 2;
				}
			}
		} else if(!seguro && pos1) {//Caso en el que se mata sí o sí
			muerto = color1;
			color1 = s;
			ultimo = 1;
		}
		return muerto;
	}
	//Función de sacar ficha
	public void sacar(String s) {
		if(color1==s) {
			pos1 = false;
			color1 = "";
			if(puente) puente = false;
			ultimo = 2;
		}else {
			System.out.println("OUT");
			pos2 = false;
			color2 ="";
			if(puente) puente = false;
			ultimo = 1;
		}
	}
}

File: Casilla.js
class Casilla{
	constructor(b,b1,s){
		this.seguro=b
		this.salida=b1
		this.colorSalida=s
		this.puente=false
		this.pos1=false
		this.pos2=false
		this.color1=null
		this.color2=null
		this.ultimo=0
	}
	gseguro(){return this.seguro}
	gsalida(){return this.salida}
	gcolorSalida(){return this.colorSalida}
	gpuente(){return this.puente}
	gpos1(){return this.pos1}
	gpos2(){return this.pos2}
	gcolor1(){return this.color1}
	gcolor2(){return this.color2}
	gultimo(){return this.ultimo}
	sePuede(s){
		return !this.pos1 || (this.seguro && !this.pos2) || this.esSalidaSuya(s)
	}
	esValido(s){
		return (!this.pos1 || (this.pos1 && s!==this.color1) || (this.seguro && (!this.pos1 || !this.pos2)))
	}
	//Comprueba si mata
	seMata(s) {
		if(!this.seguro) {
			if(this.pos1 && s!=this.color1) {
				return true;
			}
		}return false;
	}
	esSalidaSuya(s){
		return this.salida && (s===this.colorSalida && (this.color2!==s || this.color1!==s))
	}
	sacar(s){
		if(this.color1===s){
			this.pos1 = false
			this.color1 = null
			if(this.puente) this.puente = false
			this.ultimo=2
		}else{
			this.pos2=false
			this.color2=null
			if(this.puente) this.puente = false
			this.ultimo=1
		}
	}
	introducir(s){
		let muerto = "NO"
		if(!this.pos1 && !this.pos2){
			this.pos1 = true
			this.color1 = s
			this.ultimo = 1
		}else if(this.seguro && !this.pos1){
			this.pos1 = true;
			this.color1 = s;
			this.puente = true;
			this.ultimo = 1;
		} else if(this.seguro && !this.pos2) {	//Pos2 vacía, pero al ser seguro crea puente
			this.pos2 = true;
			this.color2 = s;
			this.puente = true;
			this.ultimo = 2;
		} else if(this.esSalidaSuya(s)) {	//Caso en el que es casilla salida
			if(this.ultimo===1 && s!==this.color1) {
				muerto = this.color1;
				this.color1 = s;
			}else if (this.ultimo===2 && s!==this.color2) {
				muerto = this.color2;
				this.color1 = s;
			}else { // Creo que este else sobraria
				if(s!==this.color1) {
					muerto = this.color1;
					this.color1 = s;			
					this.ultimo = 1;
				}else {
					muerto = this.color2;
					this.color2 = s;
					this.ultimo = 2;
				}
			}
			this.puente = true;

		} else if(!this.seguro && this.pos1) {//Caso en el que se mata sí o sí
			muerto = this.color1;
			this.color1 = s;
			this.ultimo = 1;
		} 
		return muerto;
	}
}

module.exports = Casilla